File: app/backtesting/backtest_engine.py
from app.backtesting.metrics.performance_metrics import (
    PerformanceMetrics
)

from app.core.logger import logger


class BacktestEngine:

    def __init__(
        self,
        strategy,
        initial_balance=10000
    ):

        self.strategy = strategy

        self.initial_balance = initial_balance

        self.balance = initial_balance

        self.position = None

        self.trade_history = []

        self.equity_curve = []

    def run(
        self,
        dataframe
    ):

        dataframe = (
            self.strategy.generate_signals(
                dataframe
            )
        )

        previous_signal = 0

        for _, row in dataframe.iterrows():

            signal = row["signal"]

            price = row["close"]

            if (
                signal == 1
                and previous_signal != 1
            ):

                self.position = price

                logger.info(
                    f"BUY at {price}"
                )

            elif (
                signal == -1
                and self.position
            ):

                pnl = (
                    price
                    - self.position
                )

                self.balance += pnl

                self.trade_history.append(
                    pnl
                )

                logger.info(
                    f"SELL at {price} "
                    f"PnL={pnl}"
                )

                self.position = None

            self.equity_curve.append(
                self.balance
            )

            previous_signal = signal

        total_return = (
            PerformanceMetrics
            .calculate_total_return(
                self.initial_balance,
                self.balance
            )
        )

        sharpe_ratio = (
            PerformanceMetrics
            .calculate_sharpe_ratio(
                self.trade_history
            )
        )

        max_drawdown = (
            PerformanceMetrics
            .calculate_max_drawdown(
                self.equity_curve
            )
        )

        return {
            "initial_balance":
                self.initial_balance,

            "final_balance":
                self.balance,

            "total_return_percent":
                total_return,

            "total_trades":
                len(self.trade_history),

            "sharpe_ratio":
                sharpe_ratio,

            "max_drawdown_percent":
                max_drawdown
        }

File: backend/src/backtesting/backtestEngine.ts
import { CSVRow } from "./data/csvLoader";
import { PerformanceMetrics } from "./metrics/performanceMetrics";
import { logger } from "../core/logger";

export class BacktestEngine {
  private strategy: any;
  private initialBalance: number;
  private balance: number;
  private position: number | null = null;
  private tradeHistory: number[] = [];
  private equityCurve: number[] = [];

  constructor(strategy: any, initialBalance: number = 10000) {
    this.strategy = strategy;
    this.initialBalance = initialBalance;
    this.balance = initialBalance;
  }

  run(dataframe: CSVRow[]) {
    const data = this.strategy.generateSignals(dataframe);
    let previousSignal = 0;

    for (const row of data) {
      const signal = row.signal;
      const price = row.close;

      if (signal === 1 && previousSignal !== 1) {
        this.position = price;
        logger.info(`BUY at ${price}`);
      } else if (signal === -1 && this.position !== null) {
        const pnl = price - this.position;
        this.balance += pnl;
        this.tradeHistory.push(pnl);
        logger.info(`SELL at ${price} PnL=${pnl}`);
        this.position = null;
      }

      this.equityCurve.push(this.balance);
      previousSignal = signal;
    }

    const totalReturn = PerformanceMetrics.calculateTotalReturn(this.initialBalance, this.balance);
    const sharpeRatio = PerformanceMetrics.calculateSharpeRatio(this.tradeHistory);
    const maxDrawdown = PerformanceMetrics.calculateMaxDrawdown(this.equityCurve);

    return {
      initial_balance: this.initialBalance,
      final_balance: this.balance,
      total_return_percent: totalReturn,
      total_trades: this.tradeHistory.length,
      sharpe_ratio: sharpeRatio,
      max_drawdown_percent: maxDrawdown,
    };
  }
}
export default BacktestEngine;
